documentation/projects/proposals/trust_and_safety/content_report_moderation/20231208-implementation_plan_django_admin_moderator_access.md

```diff
@@ -0,0 +1,468 @@
+
+There are two broad aspects of this implementation plan:
+
+1. Access control for moderator users (in other words, users who should only
+   have access to moderation tools and nothing else)
+1. Improvements and additions to Django admin tools for the content report
+   models
+
+There is also a significant bug that needs addressing: If a media object has
+multiple sensitivity reports, only one can be saved as confirmed because there
+is a
+[one-to-one relationship between the `AbstractMatureMedia` and `AbstractMedia` models](https://github.com/WordPress/openverse/blob/aa16d4f1be7607b12c428886b9890bdd947cc71c/api/api/models/media.py#L346-L355)
+and the report model admin does not handle this. The likelihood of this should
+be all but eliminated by the process and tooling changes in this implementation
+plan (because moderation reports will be handled in a single unified view for a
+given work) but we'll solve this bug by just not creating the mature media
+object on save.
+
+### Additional tools and improvements
+
+We will add the following additional tools and improvements to the report admin:
+
+- `ModerationDecision`
+- Improve the table view's default behaviour and sorting
+- A media moderation admin view to centralise moderation decisions for a given
+  work to include all pending and historical reports and show the work's
+  metadata
+
+#### `ModerationDecision`
+
+We will create a new `ModerationDecision` model for each media type, to
+accompany the existing report models. The decision model has a many-to-one
+relationship with the _media_ and tracks the history of moderation decisions
+made for the media, including explanatory notes from moderators. It is not
+possible to use `LogEntry`, like the project plan suggests to do, because
+`LogEntry`'s `object_id` column is not indexed, and querying it to find all the
+decisions for a given record would not scale.
+
+This table anticipates additional abilities for moderators in the future
+including reversing previous decisions (currently impossible, but needed if
+users and especially creators can contest decisions) or explicitly marking works
+as "not sensitive" (for example, if the work was sensitive because of erroneous
+sensitive text detection). Right now the data model and approach to sensitivity
+and deindexing don't make these possible. It is outside the scope of this
+project to add those features, but this new table, especially its `action`
+column, is intentionally flexible to allow for these kinds of things in the
+future.
+
+It will have the following columns:
+
+- `id` - basic auto-incrementing id
+- `moderator_id` - a foreign key relationship with moderator who made the
+  decision
+- `media_identifier` - a one-to-one relationship to the media the decision is
+  related to
+  - Because deindexing media removes them from the API database, this needs to
+    be a non-constrained column, so not a foreign key
+- `explanation` - moderator notes explaining the decision
```

Comment:
I had imagined it as required but it's a good point that it can be optional. Making it required would complicate the bluk actions stuff that the other IP (that Staci is assigned to) needs to implement.

Reply:
I've updated the text to say it isn't required in https://github.com/WordPress/openverse/pull/3494/commits/5acaa5b00c3830845200c5033e722133b659e2e4